### PR TITLE
Introduce self-deregistering closures for on / onany

### PR DESCRIPTION
**Pull request by jkrumbiegel**
We still have the problem that it's hard to keep track of the connections we make with Observables and to clean them up after use. This PR seeks to help while staying close enough to the previous API, that it hopefully doesn't break too much code using Observables.

The main change is that `on` returns a new type `ObserverFunction`. This wraps the closure `Function` which is normally returned by `on`. It stores the function, the observable that is being listened to, and a `weak` flag. If the `weak` flag is set via `on(..., weak = true)`, the `ObserverFunction` will call `off(observable, closure)` when finalized.

This allows for some easy housekeeping. Imagine some large plotting object that creates multiple connections via `on`. We just have to `push!` all weak `ObserverFunction`s into one big vector, and if our plot object goes out of scope, so will that vector and all the connections we made will be released.

The `weak` flag is set to `false` by default, which means that the behavior doesn't change from what we have now.

Even with `weak = false`, this PR can make it a bit easier to clean up Observables, as the `ObserverFunctions` carry both the closure and related Observable, and it's therefore easier to call `off(observerfunc)` without having to find both of these parts first.

I see almost no code storing the result of `on` in the wild right now, which would be necessary to do housekeeping (which is not being done). It would therefore be easy with this PR to improve code by storing the result and calling `off` on it at an appropriate time, or making the whole thing `weak` and let scoping handle it. If preexisting code is storing the result of `on`, then the receiving structure is probably typed `Function`, and as I've made `ObserverFunction <: Function` this should also not be breaking.

CI is currently failing because of some problem with ObservablePairs. I have never used this and cannot understand from the code what they are supposed to do. Neither are they documented. So I assume they are not very important.

### CODE REVIEW

**Comment by jkrumbiegel** [comment]
@SimonDanisch @piever I fixed the problem with observable pairs, it was just an example of code that actually _did_ do something with the return value of `on`.

Do you see any other potential problems with this change? Otherwise it would be great to add this soon, so I can use it to clean up some code in MakieLayout.

**Review by SimonDanisch — APPROVED** [review]

**Comment by SimonDanisch** [comment]
Ok, I think there should be no problem merging this! @shashi, do you have any comments?

File: src/Observables.jl [hunk 1]
```diff
@@ -84,18 +84,63 @@ end
...
+    weak::Bool
+
+    function ObserverFunction(f, observable::AbstractObservable, weak)
+        obsfunc = new(f, observable, weak)
+
+        # If the weak flag is set, deregister the function f from the observable
+        # storing it in its listeners once the ObserverFunction is garbage collected.
+        # This should free all resources associated with f unless there
+        # is another reference to it somewhere else.
+        finalizer(obsfunc) do obsfunc
```
Comment (shashi) [inline src/Observables.jl:117]: if you don't intend for the `weak` field to be changed at runtime, maybe you can move this if condition outside...
```suggestion
        finalizer(obsfunc) do obsfunc
```

File: test/runtests.jl [hunk 2]
```diff
@@ -44,6 +44,72 @@ end
...
+    @test length(Observables.listeners(a)) == 1
+    @test finalized_flag[] == false
+
+    GC.enable(true)
+    GC.gc()
+    # somehow this needs a double sweep, maybe first obsfunc, then ToFinalize?
+    GC.gc()
+    @test isempty(Observables.listeners(a))
+    @test finalized_flag[] == true
+end
```
Comment (shashi) [inline test/runtests.jl:111]: Nice test!

**Comment by shashi** [comment]
Looks good to me!

**Comment by jkrumbiegel** [comment]
so because of an edge case where the user could call `off` on a weak `ObserverFunc` which would cause the finalizer to fail later, I have changed `off` so that it doesn't return a keyerror. Just true or false depending on if it could find the function and remove it. I think that should be an acceptable change and it leaves the implementation very simple. `off` can just run again and does no harm

File: src/Observables.jl [hunk 3]
```diff
@@ -130,6 +130,8 @@ end
 
 Adds function `f` as listener to `observable`. Whenever `observable`'s value
 is set via `observable[] = val` `f` is called with `val`.
+
+Returns `true` if `f` could be removed, otherwise `false`.
```
Comment (shashi) [inline src/Observables.jl:134]: This is incorrect?

**Comment by shashi** [comment]
That change to `off` sounds good!

File: src/Observables.jl (hunk 3 above)
Comment (shashi) [reply src/Observables.jl:134]: Is this correct?

Comment (jkrumbiegel) [reply src/Observables.jl]: wow I stared at this for too long until I understood what was wrong :) good call

Comment (shashi) [reply src/Observables.jl]: Haha nice fix!

**Review by shashi — APPROVED** [review]
